### PR TITLE
Tests for DefaultConnectionService

### PR DESCRIPTION
**Pull request by gitaumoses4**
## What does this PR do?
- Adds tests for the `DefaultConnectionService` class.

## Description of tasks to be completed.
- [x] Use robolectric to test the DefaultConnectionService

## How to test this PR?
- Run the test files that were created.

## Any addition comments.
N/A

### CODE REVIEW

File: common/src/test/java/com/microsoft/identity/common/adal/internal/PowerManagerWrapperShadow.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import android.content.Context;
+
+import org.robolectric.annotation.Implementation;
+import org.robolectric.annotation.Implements;
+
+@Implements(PowerManagerWrapper.class)
+public class PowerManagerWrapperShadow {
+
+    public static boolean deviceIdleMode = true;
+    public static boolean ignoringBatteryOptimizations = true;
```
Comment (shahzaibj) [inline common/src/test/java/com/microsoft/identity/common/adal/internal/PowerManagerWrapperShadow.java:35]: I don't see any reason for these to be static...in fact I don't see any reason for these to even be defined (based on their current usage)

File: common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    protected Context mContext;
+
+    @Before
+    public void setup() {
+        mContext = ApplicationProvider.getApplicationContext();
+    }
+
+    @Test
+    public void testConnectionWhenConnected() {
+        DefaultConnectionService defaultConnectionService = new DefaultConnectionService(mContext);
```
Comment (shahzaibj) [inline common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java:60]: nit: final here and everywhere

**Comment by shahzaibj** [comment]
Why does the following test (circled) take 7 seconds to run? Is this an indication of a problem?

image

**Comment by gitaumoses4** [comment]
> Why does the following test (circled) take 7 seconds to run? Is this an indication of a problem?
> 
> image

Screenshot from 2021-05-26 20-27-24

@shahzaibj Seems the test that's run first always takes longer. I just changed the name so that the one you marked was run last.

File: common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    protected Context mContext;
+
+    @Before
+    public void setup() {
+        mContext = ApplicationProvider.getApplicationContext();
+    }
+
+    @Test
+    public void testConnectionWhenConnected() {
```
Comment (shahzaibj) [inline common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java:59]:
```suggestion
    public void testConnectionAvailableWhenConnected() {
```

File: common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        final DefaultConnectionService defaultConnectionService = new DefaultConnectionService(mContext);
+
+        assertTrue(defaultConnectionService.isConnectionAvailable());
+    }
+
+    @Test
+    public void testConnectionWhenNoActiveNetwork() {
```
Comment (shahzaibj) [inline common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java:66]:
```suggestion
    public void testConnectionNotAvailableWhenNoActiveNetwork() {
```

File: common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    @Test
+    public void testNetworkDisabledFromOptimizations() {
+        final DefaultConnectionService defaultConnectionService = new DefaultConnectionService(mContext);
+        ReflectionHelpers.setStaticField(Build.VERSION.class, "SDK_INT", Build.VERSION_CODES.M);
+
+        assertTrue(defaultConnectionService.isNetworkDisabledFromOptimizations());
+    }
+
+    @Test
+    public void testNetworkDisabledFromOptimizationsForAPIsBelow23() {
```
Comment (shahzaibj) [inline common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java:85]:
```suggestion
    public void testNetworkNotDisabledFromOptimizationsForAPIsBelow23() {
```

File: common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                .getSystemService(Context.CONNECTIVITY_SERVICE);
+
+        shadowOf(connectivityManager).setActiveNetworkInfo(null);
+        final DefaultConnectionService defaultConnectionService = new DefaultConnectionService(mContext);
+
+        assertFalse(defaultConnectionService.isConnectionAvailable());
+    }
+
+    @Test
+    public void testNetworkDisabledFromOptimizations() {
```
Comment (shahzaibj) [inline common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java:77]:
```suggestion
    public void testNetworkIsDisabledFromOptimizationsForAPIs23AndAbove() {
```

**Review by shahzaibj — APPROVED** [review]
Left some suggestions..else lgtm

File: common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    @Test
+    public void testConnectionNotAvailableWhenNoActiveNetwork() {
+        final ConnectivityManager connectivityManager = (ConnectivityManager) mContext
+                .getSystemService(Context.CONNECTIVITY_SERVICE);
+
+        shadowOf(connectivityManager).setActiveNetworkInfo(null);
+        final DefaultConnectionService defaultConnectionService = new DefaultConnectionService(mContext);
+
+        assertFalse(defaultConnectionService.isConnectionAvailable());
+    }
```
Comment (AdamBJohnsonx) [inline common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java:74]: What I would suggest here is to trigger a few different network events, if possible.  If a single network is unavailable, we should still be able to connect, or if we lose one connection and gain another we should remain connected.

Comment (AdamBJohnsonx) [reply common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java:74]: I think the following permutations would be sufficient:

Have 2, lose 1, stay connected.
Have 2, lose both, not connected, gain 1, stay connected.
Have none, not connected, gain 1, are connected.
Have none, gain 2, are connected, lose 1, are connected.

The idea being that we can then bring back the changes to be event driven and have these tests tell us that they're safe.

Comment (gitaumoses4) [reply common/src/test/java/com/microsoft/identity/common/adal/internal/net/DefaultConnectionServiceTest.java:74]: Brilliant!! ... Thanks Adam.

**Review by AdamBJohnsonx — APPROVED** [review]

**Pull request merged**